### PR TITLE
Testcases related to starkgate bridging

### PR DESCRIPTION
**Pull request by anshalshukla**
Adds testcases for both eth and erc20 token bridging btw eth and madara

# Pull Request type

- Testing

## Does this introduce a breaking change?

No

## Other information

Dependent on Zaun [PR#9](https://github.com/keep-starknet-strange/zaun/pull/9)

### CODE REVIEW

File: starknet-e2e-test/src/starknet_sovereign.rs [hunk 1]
```diff
@@ -47,11 +47,14 @@ impl StarknetSovereign {
     pub async fn deploy() -> Self {
         // Try to attach to an already running sandbox (GitHub CI case)
         // otherwise spawn new sandbox instance
-        let sandbox = if let Ok(endpoint) = std::env::var("ANVIL_ENDPOINT") {
-            EthereumSandbox::attach(Some(endpoint)).expect("Failed to attach to sandbox")
-        } else {
-            EthereumSandbox::spawn(None)
-        };
+        // let sandbox = if let Ok(endpoint) = std::env::var("ANVIL_ENDPOINT") {
```
Comment (EvolveArt) [inline starknet-e2e-test/src/starknet_sovereign.rs:50]: remove code

File: starknet-e2e-test/src/token_bridge.rs [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+        self.registry.client()
+    }
+    pub fn bridge_client(&self) -> Arc<LocalWalletSignerMiddleware> {
+        self.token_bridge.client()
+    }
+    pub fn dai_client(&self) -> Arc<LocalWalletSignerMiddleware> {
+        self.dai_erc20.client()
+    }
+
+    pub async fn deploy(client: Arc<LocalWalletSignerMiddleware>) -> Self {
```
Comment (EvolveArt) [inline starknet-e2e-test/src/token_bridge.rs:65]: let's define an `Deployable` trait, wdyt ?

File: starknet-e2e-test/src/utils.rs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    );
+
+    madara_write_lock
+        .create_block_with_txs(vec![Transaction::Execution(Execution::from(deploy_tx))])
+        .await
+        .expect("Unable to deploy ERC20 token on madara");
+    deploy_tx.deployed_address()
+}
+
+pub async fn madara_contract_call(
```
Comment (EvolveArt) [inline starknet-e2e-test/src/utils.rs:54]: name is misleading maybe `invoke_contract` is better ?

File: starknet-e2e-test/src/utils.rs [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+use starknet_rpc_test::utils::{build_single_owner_account, AccountActions};
+use starknet_rpc_test::Transaction;
+use tokio::time::sleep;
+
+pub async fn deploy_erc20_token_on_l2(madara: &ThreadSafeMadaraClient, minter: FieldElement) -> FieldElement {
+    let rpc = madara.get_starknet_client().await;
+    let account = build_single_owner_account(&rpc, SIGNER_PRIVATE, CAIRO_1_ACCOUNT_CONTRACT, false);
+
+    let (declare_tx, class_hash, _) = account.declare_contract(
+        "../starknet-e2e-test/contracts/erc20.sierra.json",
```
Comment (EvolveArt) [inline starknet-e2e-test/src/utils.rs:18]: declare these as constants

File: starknet-e2e-test/src/token_bridge.rs [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+            .await
+            .expect("Failed to approve dai transfer for Starknet Token Bridge");
+    }
+
+    pub async fn token_balance(&self, address: Address) -> U256 {
+        self.dai_erc20.balance_of(address).await.unwrap()
+    }
+}
+
+fn pad_bytes(address: Address) -> Vec<u8> {
```
Comment (EvolveArt) [inline starknet-e2e-test/src/token_bridge.rs:264]: move to utils

File: starknet-e2e-test/ethereum_e2e.rs [hunk 6]
```diff
@@ -44,3 +57,142 @@ async fn madara_advances_ethereum_settlement_contract_state_in_sovereign_mode()
...
+    let mut madara_write_lock = madara.write().await;
+    madara_write_lock.create_n_blocks(2).await.expect("Unable to create empty blocks in madara");
+    sleep(Duration::from_millis(60000)).await;
+
+    let l1_recipient: Address = Address::from_str("0x59FA6981892396D67BBd31e80E9d91506213335F").unwrap();
+    let balance_before = token_bridge.token_balance(l1_recipient).await;
+    token_bridge.withdraw(token_bridge.dai_address(), 5.into(), l1_recipient).await;
+    let balance_after = token_bridge.token_balance(l1_recipient).await;
+
+    assert_eq!(balance_before + U256::from_dec_str("5").unwrap(), balance_after);
```
Comment (EvolveArt) [inline starknet-e2e-test/ethereum_e2e.rs:195]: not clear where the `5` comes from, please define it as a constant

File: starknet-e2e-test/ethereum_e2e.rs [hunk 7]
```diff
@@ -44,3 +57,142 @@ async fn madara_advances_ethereum_settlement_contract_state_in_sovereign_mode()
...
+    starknet_sovereign.initialize_for_goerli(0u64.into(), 0u64.into()).await;
+
+    let token_bridge = StarknetTokenBridge::deploy(starknet_sovereign.client().clone()).await;
+    let l2_bridge_address = StarknetTokenBridge::deploy_l2_contracts(&madara).await;
+
+    token_bridge.initialize(starknet_sovereign.address()).await;
+    token_bridge.setup_l2_bridge(madara, l2_bridge_address).await;
+    token_bridge
+        .setup_l1_bridge(
+            H160::from_str("0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266").unwrap(),
```
Comment (EvolveArt) [inline starknet-e2e-test/ethereum_e2e.rs:125]: declare all addresses as constants

File: starknet-e2e-test/ethereum_e2e.rs [hunk 8]
```diff
@@ -44,3 +57,142 @@ async fn madara_advances_ethereum_settlement_contract_state_in_sovereign_mode()
 
     Ok(())
 }
+
+#[rstest]
+#[tokio::test]
+async fn eth_bridge(madara: &ThreadSafeMadaraClient) -> Result<(), anyhow::Error> {
```
Comment (EvolveArt) [inline starknet-e2e-test/ethereum_e2e.rs:63]:
```suggestion
async fn deposit_and_withdraw_from_eth_bridge(madara: &ThreadSafeMadaraClient) -> Result<(), anyhow::Error> {
```

File: starknet-e2e-test/ethereum_e2e.rs [hunk 9]
```diff
@@ -44,3 +57,142 @@ async fn madara_advances_ethereum_settlement_contract_state_in_sovereign_mode()
...
+    let balance_after = eth_bridge.eth_balance(l1_recipient).await;
+
+    assert_eq!(balance_before + U256::from_dec_str("5").unwrap(), balance_after);
+
+    Ok(())
+}
+
+#[rstest]
+#[tokio::test]
+async fn erc20_bridge(madara: &ThreadSafeMadaraClient) -> Result<(), anyhow::Error> {
```
Comment (EvolveArt) [inline starknet-e2e-test/ethereum_e2e.rs:114]:
```suggestion
async fn deposit_and_withdraw_erc20_bridge(madara: &ThreadSafeMadaraClient) -> Result<(), anyhow::Error> {
```

File: starknet-e2e-test/ethereum_e2e.rs [hunk 10]
```diff
@@ -44,3 +57,142 @@ async fn madara_advances_ethereum_settlement_contract_state_in_sovereign_mode()
...
+
+#[rstest]
+#[tokio::test]
+async fn eth_bridge(madara: &ThreadSafeMadaraClient) -> Result<(), anyhow::Error> {
+    let starknet_sovereign = StarknetSovereign::deploy().await;
+    starknet_sovereign.initialize_for_goerli(0u64.into(), 0u64.into()).await;
+
+    let eth_bridge = StarknetLegacyEthBridge::deploy(starknet_sovereign.client().clone()).await;
+    let l2_bridge_address = StarknetLegacyEthBridge::deploy_l2_contracts(&madara).await;
+    let erc20_address = deploy_erc20_token_on_l2(&madara, l2_bridge_address).await;
```
Comment (apoorvsadana) [inline starknet-e2e-test/ethereum_e2e.rs:69]:
```suggestion
    let eth_address = deploy_erc20_token_on_l2(&madara, l2_bridge_address).await;
```

File: starknet-e2e-test/src/eth_bridge.rs [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+        madara_write_lock
+            .create_block_with_txs(vec![Transaction::Execution(Execution::from(deploy_tx))])
+            .await
+            .expect("Unable to deploy legacy token bridge on l2");
+        deploy_tx.deployed_address()
+    }
+
+    /// Initialize Starknet core contract with the specified data.
+    ///
+    /// Also register Anvil default account as an operator.
```
Comment (apoorvsadana) [inline starknet-e2e-test/src/eth_bridge.rs:71]: wrong comment?

File: starknet-e2e-test/src/token_bridge.rs [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+
+        let mut txs = madara_write_lock.create_block_with_txs(vec![Transaction::Execution(deploy_tx)]).await.unwrap();
+
+        let deploy_tx_result = txs.pop().unwrap();
+        get_contract_address_from_deploy_tx(&rpc, deploy_tx_result).await.unwrap()
+    }
+
+    /// Initialize Starknet core contract with the specified data.
+    ///
+    /// Also register Anvil default account as an operator.
```
Comment (apoorvsadana) [inline starknet-e2e-test/src/token_bridge.rs:129]: wrong comment?

**Comment by anshalshukla** [comment]
@apoorvsadana @tdelabro not sure why some of the test cases are failing because I didn't make any modifications in them as such. Also, right now in cargo.toml I've updated the zaun to point to my repo but I think we want to get that PR merged first and then update the version here. 

Please re review this PR as well and [here](https://github.com/keep-starknet-strange/zaun/pull/9) is the link to zaun PR.

File: crates/client/data-availability/Cargo.toml [hunk 13]
```diff
@@ -57,7 +57,15 @@ mp-hashers = { workspace = true, default-features = true }
...
 # Zaun
+ethereum-instance = { workspace = true }
+starkgate-manager-client = { workspace = true }
+starkgate-registry-client = { workspace = true }
 starknet-core-contract-client = { workspace = true }
+starknet-erc20-client = { workspace = true }
+starknet-eth-bridge-client = { workspace = true }
+starknet-proxy-client = { workspace = true }
+starknet-token-bridge-client = { workspace = true }
+zaun-utils = { workspace = true }
```
Comment (apoorvsadana) [inline crates/client/data-availability/Cargo.toml:68]: do we need this @ocdbytes?

File: crates/client/l1-messages/Cargo.toml [hunk 14]
```diff
@@ -40,7 +40,15 @@ sc-client-api = { workspace = true, default-features = true }
...
 # Zaun
+ethereum-instance = { workspace = true }
+starkgate-manager-client = { workspace = true }
+starkgate-registry-client = { workspace = true }
 starknet-core-contract-client = { workspace = true }
+starknet-erc20-client = { workspace = true }
+starknet-eth-bridge-client = { workspace = true }
+starknet-proxy-client = { workspace = true }
+starknet-token-bridge-client = { workspace = true }
+zaun-utils = { workspace = true }
```
Comment (apoorvsadana) [inline crates/client/l1-messages/Cargo.toml:51]: do we need this @ocdbytes?

File: crates/client/settlement/Cargo.toml [hunk 15]
```diff
@@ -50,7 +50,15 @@ pallet-starknet-runtime-api = { workspace = true, default-features = true }
...
 # Zaun
+ethereum-instance = { workspace = true }
+starkgate-manager-client = { workspace = true }
+starkgate-registry-client = { workspace = true }
 starknet-core-contract-client = { workspace = true }
+starknet-erc20-client = { workspace = true }
+starknet-eth-bridge-client = { workspace = true }
+starknet-proxy-client = { workspace = true }
+starknet-token-bridge-client = { workspace = true }
+zaun-utils = { workspace = true }
```
Comment (apoorvsadana) [inline crates/client/settlement/Cargo.toml:61]: do we need all these @ocdbytes?

File: starknet-test-utils/Cargo.toml [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+starknet-accounts = { workspace = true }
+starknet-contract = { workspace = true }
+starknet-core = { workspace = true }
+starknet-crypto = { workspace = true }
+starknet-ff = { workspace = true }
+starknet-providers = { workspace = true }
+starknet-signers = { workspace = true }
+thiserror = { workspace = true }
+tokio = { version = "1.36.0", features = ["rt", "macros", "parking_lot"] }
+url = "2.4.1"
```
Comment (apoorvsadana) [inline starknet-test-utils/Cargo.toml:27]: let's use workspace = true wherever possible

**Review by apoorvsadana — APPROVED** [review]

**Review by EvolveArt — APPROVED** [review]

**Pull request merged**